File: backend-node/src/services/AnomalyService.ts
import { createAlert } from "./AlertService.js";
import { calculateDeviation, isPointWithinRadius } from "../utils/geoFence.js";
import { getActiveRiskZones, updateTourist, type IRiskZone } from "./mongoStore.js";
import type { ITourist } from "../schemas/index.js";

const INACTIVITY_THRESHOLD_MINUTES = 30;
const DEVIATION_THRESHOLD_KM = 5.0;

const touristActiveZones = new Map<string, Set<number>>();

export async function processLocation(tourist: ITourist, accuracy?: number) {
  await checkInactivity(tourist);
  await checkRouteDeviation(tourist);
  await checkGeoFence(tourist);
}

async function checkInactivity(tourist: ITourist) {
  if (!tourist.lastSeen) {
    return;
  }
  const lastSeen = new Date(tourist.lastSeen).getTime();
  const minutesSinceLastSeen = (Date.now() - lastSeen) / 60000;
  if (minutesSinceLastSeen > INACTIVITY_THRESHOLD_MINUTES) {
    await createAlert({
      touristId: tourist._id,
      alertType: "INACTIVITY",
      latitude: tourist.currentLat,
      longitude: tourist.currentLng,
      message: `Tourist has not sent a location update in ${Math.floor(minutesSinceLastSeen)} minutes.`
    });
  }
}

async function checkRouteDeviation(tourist: ITourist) {
  const deviationKm = calculateDeviation(tourist.currentLat, tourist.currentLng);
  if (deviationKm > DEVIATION_THRESHOLD_KM) {
    await createAlert({
      touristId: tourist._id,
      alertType: "DEVIATION",
      latitude: tourist.currentLat,
      longitude: tourist.currentLng,
      message: `Route deviation detected: ${deviationKm.toFixed(2)} km off planned route.`
    });
  }
}

async function checkGeoFence(tourist: ITourist) {
  const { currentLat: lat, currentLng: lng } = tourist;
  if (typeof lat !== "number" || typeof lng !== "number") {
    return;
  }

  const activeZones = await getActiveRiskZones();
  if (activeZones.length === 0) {
    touristActiveZones.delete(tourist._id);
    return;
  }

  const zoneLookup = new Map(activeZones.map((zone) => [zone.zoneId, zone]));

  const currentlyInside = new Set<number>(
    activeZones
      .filter((zone) => isPointWithinRadius(lat, lng, zone.centerLat, zone.centerLng, zone.radiusMeters))
      .map((zone) => zone.zoneId)
  );

  const previous = touristActiveZones.get(tourist._id) ?? new Set<number>();
  const entered = new Set<number>([...currentlyInside].filter((id) => !previous.has(id)));

  if (currentlyInside.size > 0) {
    touristActiveZones.set(tourist._id, currentlyInside);
  } else {
    touristActiveZones.delete(tourist._id);
  }

  if (entered.size === 0) {
    return;
  }

  let safetyScore = tourist.safetyScore ?? 100;
  for (const zoneId of entered) {
    const zone = zoneLookup.get(zoneId);
    if (!zone) {
      continue;
    }
    safetyScore = Math.max(0, safetyScore - penaltyFor(zone));
    await createAlert({
      touristId: tourist._id,
      alertType: "RISK_ZONE",
      latitude: lat,
      longitude: lng,
      message: `Tourist entered risk zone '${zone.name}' [${zone.riskLevel}]`
    });
  }
  // Update the tourist's safety score in DB
  await updateTourist(tourist._id, { safetyScore: Math.max(0, Math.min(100, safetyScore)) });
}

function penaltyFor(zone: IRiskZone) {
  if (!zone.riskLevel) {
    return 8.0;
  }
  switch (zone.riskLevel) {
    case "LOW":
      return 5.0;
    case "MEDIUM":
      return 10.0;
    case "HIGH":
      return 18.0;
    default:
      return 8.0;
  }
}

File: backend-spring/src/main/java/com/safarsathi/service/AnomalyService.java
package com.safarsathi.service;

import com.safarsathi.entity.Alert;
import com.safarsathi.entity.RiskZone;
import com.safarsathi.entity.Tourist;
import com.safarsathi.repository.TouristRepository;
import com.safarsathi.util.GeoFenceUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * Detects anomalies: inactivity, route deviation, geo-fence breaches.
 */
@Service
@RequiredArgsConstructor
public class AnomalyService {

    private static final int INACTIVITY_THRESHOLD_MINUTES = 30;
    private static final double DEVIATION_THRESHOLD_KM = 5.0;

    private final AlertService alertService;
    private final RiskZoneService riskZoneService;
    private final TouristRepository touristRepository;

    private final Map<String, Set<Integer>> touristActiveZones = new ConcurrentHashMap<>();

    public void processLocation(Tourist tourist) {
        checkInactivity(tourist);
        checkRouteDeviation(tourist);
        checkGeoFence(tourist);
    }

    private void checkInactivity(Tourist tourist) {
        if (tourist.getLastSeen() == null) return;
        try {
            long lastSeenMs = java.time.Instant.parse(tourist.getLastSeen()).toEpochMilli();
            double minutesSince = (System.currentTimeMillis() - lastSeenMs) / 60000.0;
            if (minutesSince > INACTIVITY_THRESHOLD_MINUTES) {
                Alert alert = Alert.builder()
                        .touristId(tourist.getId())
                        .alertType("INACTIVITY")
                        .latitude(tourist.getCurrentLat())
                        .longitude(tourist.getCurrentLng())
                        .message(String.format("Tourist has not sent a location update in %d minutes.",
                                (int) minutesSince))
                        .build();
                alertService.createAlert(alert);
            }
        } catch (Exception ignored) {
            // Invalid timestamp, skip
        }
    }

    private void checkRouteDeviation(Tourist tourist) {
        double deviationKm = GeoFenceUtil.calculateDeviation(tourist.getCurrentLat(), tourist.getCurrentLng());
        if (deviationKm > DEVIATION_THRESHOLD_KM) {
            Alert alert = Alert.builder()
                    .touristId(tourist.getId())
                    .alertType("DEVIATION")
                    .latitude(tourist.getCurrentLat())
                    .longitude(tourist.getCurrentLng())
                    .message(String.format("Route deviation detected: %.2f km off planned route.", deviationKm))
                    .build();
            alertService.createAlert(alert);
        }
    }

    private void checkGeoFence(Tourist tourist) {
        Double lat = tourist.getCurrentLat();
        Double lng = tourist.getCurrentLng();
        if (lat == null || lng == null) return;

        List<RiskZone> activeZones = riskZoneService.listActiveRiskZones();
        if (activeZones.isEmpty()) {
            touristActiveZones.remove(tourist.getId());
            return;
        }

        Map<Integer, RiskZone> zoneLookup = activeZones.stream()
                .collect(Collectors.toMap(RiskZone::getZoneId, z -> z));

        Set<Integer> currentlyInside = activeZones.stream()
                .filter(zone -> GeoFenceUtil.isPointWithinRadius(lat, lng,
                        zone.getCenterLat(), zone.getCenterLng(), zone.getRadiusMeters()))
                .map(RiskZone::getZoneId)
                .collect(Collectors.toSet());

        Set<Integer> previous = touristActiveZones.getOrDefault(tourist.getId(), new HashSet<>());

        Set<Integer> entered = new HashSet<>(currentlyInside);
        entered.removeAll(previous);

        if (!currentlyInside.isEmpty()) {
            touristActiveZones.put(tourist.getId(), currentlyInside);
        } else {
            touristActiveZones.remove(tourist.getId());
        }

        if (entered.isEmpty()) return;

        double safetyScore = tourist.getSafetyScore() != null ? tourist.getSafetyScore() : 100.0;
        for (Integer zoneId : entered) {
            RiskZone zone = zoneLookup.get(zoneId);
            if (zone == null) continue;
            safetyScore = Math.max(0, safetyScore - penaltyFor(zone));
            Alert alert = Alert.builder()
                    .touristId(tourist.getId())
                    .alertType("RISK_ZONE")
                    .latitude(lat)
                    .longitude(lng)
                    .message(String.format("Tourist entered risk zone '%s' [%s]",
                            zone.getName(), zone.getRiskLevel()))
                    .build();
            alertService.createAlert(alert);
        }

        tourist.setSafetyScore(Math.max(0, Math.min(100, safetyScore)));
        touristRepository.save(tourist);
    }

    private double penaltyFor(RiskZone zone) {
        if (zone.getRiskLevel() == null) return 8.0;
        return switch (zone.getRiskLevel()) {
            case "LOW" -> 5.0;
            case "MEDIUM" -> 10.0;
            case "HIGH" -> 18.0;
            default -> 8.0;
        };
    }
}
